File: pkg/tracing/tracing.go
// Package tracing wires each service into a single OpenTelemetry
// TracerProvider exporting to Jaeger over OTLP/HTTP, and provides a
// kafka.Header carrier so trace context survives the hop across Kafka —
// otel's built-in propagators only know how to inject/extract HTTP-style
// text map carriers, so producers/consumers need an adapter.
package tracing

import (
	"context"
	"fmt"
	"net/url"
	"os"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.43.0"
	"go.opentelemetry.io/otel/trace"

	kafka "github.com/segmentio/kafka-go"
)

// Init configures the global TracerProvider and propagator for the calling
// service. It reads OTEL_EXPORTER_OTLP_ENDPOINT (defaulting to
// http://localhost:4318, matching .env.example) and never fails startup:
// if the exporter can't be built, tracing is disabled and a no-op shutdown
// is returned, since a missing collector must never take down a service
// whose job is ingesting real telemetry.
func Init(ctx context.Context, serviceName string) (shutdown func(context.Context) error, err error) {
	endpoint := os.Getenv("OTEL_EXPORTER_OTLP_ENDPOINT")
	if endpoint == "" {
		endpoint = "http://localhost:4318"
	}

	// WithEndpoint (host:port, no scheme) appends the default "/v1/traces"
	// path; WithEndpointURL takes the URL's path verbatim and does NOT
	// default it, which silently 404s against Jaeger's OTLP receiver if
	// OTEL_EXPORTER_OTLP_ENDPOINT (an OTel-standard base URL, no path) is
	// passed straight through.
	host := endpoint
	if u, parseErr := url.Parse(endpoint); parseErr == nil && u.Host != "" {
		host = u.Host
	}

	exporter, err := otlptracehttp.New(ctx,
		otlptracehttp.WithEndpoint(host),
		otlptracehttp.WithInsecure(),
	)
	if err != nil {
		return func(context.Context) error { return nil }, fmt.Errorf("tracing: build otlp exporter: %w", err)
	}

	res, err := resource.New(ctx,
		resource.WithAttributes(
			semconv.ServiceName(serviceName),
		),
	)
	if err != nil {
		return func(context.Context) error { return nil }, fmt.Errorf("tracing: build resource: %w", err)
	}

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter, sdktrace.WithBatchTimeout(2*time.Second)),
		sdktrace.WithResource(res),
	)

	otel.SetTracerProvider(tp)
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	))

	return tp.Shutdown, nil
}

// Tracer returns the named tracer from the global provider — a thin
// convenience so call sites don't import otel directly.
func Tracer(name string) trace.Tracer {
	return otel.Tracer(name)
}

// kafkaHeaderCarrier adapts a *[]kafka.Header to propagation.TextMapCarrier
// so the standard W3C traceparent propagator can inject into, and extract
// from, Kafka message headers exactly as it would HTTP headers.
type kafkaHeaderCarrier struct {
	headers *[]kafka.Header
}

func (c kafkaHeaderCarrier) Get(key string) string {
	for _, h := range *c.headers {
		if h.Key == key {
			return string(h.Value)
		}
	}
	return ""
}

func (c kafkaHeaderCarrier) Set(key, value string) {
	for i, h := range *c.headers {
		if h.Key == key {
			(*c.headers)[i].Value = []byte(value)
			return
		}
	}
	*c.headers = append(*c.headers, kafka.Header{Key: key, Value: []byte(value)})
}

func (c kafkaHeaderCarrier) Keys() []string {
	keys := make([]string, len(*c.headers))
	for i, h := range *c.headers {
		keys[i] = h.Key
	}
	return keys
}

// InjectKafka writes the span context carried by ctx into headers, so a
// consumer on the other side of the broker can continue the same trace.
func InjectKafka(ctx context.Context, headers *[]kafka.Header) {
	otel.GetTextMapPropagator().Inject(ctx, kafkaHeaderCarrier{headers: headers})
}

// ExtractKafka reads any propagated span context out of a consumed
// message's headers and returns a context a consumer can start child spans
// from. If no trace context is present, it returns ctx unchanged and the
// consumer's span simply becomes a new trace root.
func ExtractKafka(ctx context.Context, headers []kafka.Header) context.Context {
	return otel.GetTextMapPropagator().Extract(ctx, kafkaHeaderCarrier{headers: &headers})
}

File: shared/tracing.py
"""Wires each service into a single OpenTelemetry TracerProvider exporting
to Jaeger over OTLP/HTTP, plus a Kafka header carrier so trace context
survives the hop across Kafka — otel's built-in propagators only know how
to inject/extract HTTP-style text map carriers, so producers/consumers need
an adapter. Python port of pkg/tracing.

confluent-kafka represents message headers as a list of (str, bytes) tuples
(or None), unlike segmentio/kafka-go's []kafka.Header — the carrier below
adapts that shape instead.
"""

from __future__ import annotations

import logging
import os
from collections.abc import Callable
from typing import Any
from urllib.parse import urlparse

from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.propagate import extract, inject
from opentelemetry.propagators.textmap import Getter, Setter
from opentelemetry.sdk.resources import SERVICE_NAME, Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

_logger = logging.getLogger(__name__)

KafkaHeaders = list[tuple[str, bytes]]


def init(service_name: str) -> Callable[[], None]:
    """Configures the global TracerProvider for the calling service. Reads
    OTEL_EXPORTER_OTLP_ENDPOINT (defaulting to http://localhost:4318,
    matching .env.example) and never raises: if the exporter can't be
    built, tracing is disabled and a no-op shutdown is returned, since a
    missing collector must never take down a service whose job is
    ingesting real telemetry.
    """
    endpoint = os.environ.get("OTEL_EXPORTER_OTLP_ENDPOINT", "http://localhost:4318")

    # The exporter wants the full OTLP path; OTEL_EXPORTER_OTLP_ENDPOINT is
    # an OTel-standard *base* URL with no path, so it needs "/v1/traces"
    # appended explicitly — the exact 404-against-Jaeger gotcha the Go
    # version's comment documents, ported here as the same fix rather than
    # rediscovering it.
    parsed = urlparse(endpoint)
    traces_endpoint = endpoint
    if parsed.scheme and parsed.netloc and not parsed.path.rstrip("/").endswith("/v1/traces"):
        traces_endpoint = endpoint.rstrip("/") + "/v1/traces"

    try:
        exporter = OTLPSpanExporter(endpoint=traces_endpoint)
        provider = TracerProvider(resource=Resource.create({SERVICE_NAME: service_name}))
        provider.add_span_processor(BatchSpanProcessor(exporter, schedule_delay_millis=2000))
        trace.set_tracer_provider(provider)
        return lambda: provider.shutdown()
    except Exception as exc:  # noqa: BLE001 - startup must never fail on a missing collector
        _logger.warning("tracing disabled: %s", exc)
        return lambda: None


def tracer(name: str) -> trace.Tracer:
    """Thin convenience so call sites don't import opentelemetry directly."""
    return trace.get_tracer(name)


class _KafkaHeaderGetter(Getter[KafkaHeaders]):
    def get(self, carrier: KafkaHeaders, key: str) -> list[str] | None:
        for h_key, h_value in carrier:
            if h_key == key:
                return [h_value.decode("utf-8")]
        return None

    def keys(self, carrier: KafkaHeaders) -> list[str]:
        return [h_key for h_key, _ in carrier]


class _KafkaHeaderSetter(Setter[KafkaHeaders]):
    def set(self, carrier: KafkaHeaders, key: str, value: str) -> None:
        for i, (h_key, _) in enumerate(carrier):
            if h_key == key:
                carrier[i] = (key, value.encode("utf-8"))
                return
        carrier.append((key, value.encode("utf-8")))


_getter = _KafkaHeaderGetter()
_setter = _KafkaHeaderSetter()


def inject_kafka(headers: KafkaHeaders) -> None:
    """Writes the current span context into headers (in place), so a
    consumer on the other side of the broker can continue the same trace."""
    inject(headers, setter=_setter)


def extract_kafka(headers: KafkaHeaders | None) -> Any:
    """Reads any propagated span context out of a consumed message's
    headers and returns an OTel context a consumer can start child spans
    from. If no trace context is present, the consumer's span simply
    becomes a new trace root."""
    return extract(headers or [], getter=_getter)
